### DISCUSSION TITLE
How to bind to property of a class instance

### DISCUSSION THREAD

**Discussion by FraserElectronics** [AvaloniaUI/Avalonia#13286] [Q&A] [answered]
I have a view whose DataContext points to the view model.
In the view model, I have an ObservableProperty which is an instance of a class.
In my view I would like to bind to a property of the class but I get an "Unable to resolve..." error.

Example View:

```
<ComboBox SelectedIndex="{TemplateBinding TheClass.SelectedIndex, Mode=TwoWay}">
```

Example View Model:

```
[ObservableProperty]
public MyClass theClass = new MyClass();
```

In this case, I get the error:

```
Unable to resolve property or method of name 'TheClass' on type ....
```

I am sure this should be straightforward but I am new to XAML and Avalonia.

Andy

**Comment by timunie** [comment]
```cs
[ObservableProperty]
private MyClass theClass = new MyClass();
```

iirc the community toolkit requires private field here.

Moreover: TemplateBinding is ONLY for ControlTemplates. Not DataTemplates or somewhere else.

**Reply by FraserElectronics** [reply]
@timunie 
Tim, well spotted - all my other observable properties are private and for some reason I had not noticed that I have made that one public ! 
It now works :-)

The TemplateBinding above was a typo on my part because I originally had the problem with a ControlTemplate.

Thanks for your help.

Andy

**Comment by FraserElectronics** [comment]
Sorry, disregard my last post - making it private works for a simple view/viewmodel but my original problem as I said is with a templated control. So in my templated control I have:

```
<ComboBox SelectedIndex="{Calibration.CalibrationType, Mode=TwoWay}">
```

and in the code-behind I have:

```
    public static readonly DirectProperty<CalibrationControl,ICalibrationModel?> CalibrationProperty =
    AvaloniaProperty.RegisterDirect<CalibrationControl, ICalibrationModel?>(nameof(Calibration), o => o.Calibration, ( o, v ) => o.Calibration = v );
    private ICalibrationModel? _calibration;
    public ICalibrationModel? Calibration
    {
        get => _calibration;
        set => SetAndRaise( CalibrationProperty, ref _calibration, value );
    }
```
The error is:

```
Unable to resolve type Calibration.CalibrationType from namespace...
```

The ICalibrationModel interface is in a different assembly so not sure if this is a namespace issue?

Andy

**Reply by timunie** [reply]
at least someting missing here. 
`<ComboBox SelectedIndex="{TemplateBinding Calibration.CalibrationType, Mode=TwoWay}">`

Not sure if it works that way really good. I'd need to have sample project to try on my own.

**Reply by FraserElectronics** [reply]
So basically, that class has a dozen properties, and I would like to avoid having to expose each one of them as separate DirectProperties on my TemplatedControl. The above is kind of pseudo-code as the issue is within a much larger project. I will try and put together a very simple solution that shows the issue.

Sorry for all the questions - I am a 30 year veteran of WinForms programming but kind of skipped over the whole WPF thing as I never had the time to really learn how to use it. I do see Avalonia as very attractive for its multi-platform capability especially on embedded systems where I do a lot or work.

Andy

**Reply by timunie** [reply]
You're still welcome, but also I hope you understand we can't guess all what's going on. So I really need a sample here. 

Or as an alternative you need to get paid support, and we can help you inside your larger project with questions. But I cannot offer that for free.

**Reply by FraserElectronics** [reply]
@timunie  Tim, I appreciate that your time is not free - I sit on the same side of the fence with my customers :-(

I thought I was just doing something silly that would be easy to point out.
I will put together an example solution but please don't feel obliged to fault find it, I can just stick to my individual DirectProperties for now.

BR

Andy

**Reply by FraserElectronics** [reply]
I have created a repo on Github that shows the problem I am having.
I can't believe what I want to achieve cannot be done, I guess I am missing something simple.

The repo: 

https://github.com/FraserElectronics/AvaloniaClassMemberInTemplatedControl

If anyone can point out where I am going wrong it would be appreciated.

Andy

**Reply by jack775544** [reply, marked as answer]
You have 2 issues in your project.

The issue first is noted in [this document](https://docs.avaloniaui.net/docs/data-binding/binding-in-a-control-template), namely:

> TemplateBinding accepts only a single property rather than a property path

So changing the declaration for the NumericUpDown control to the following will work.

```xml
<NumericUpDown Value="{Binding Record.Value1, RelativeSource={RelativeSource TemplatedParent}}" Width="200" Background="Azure"/>
```

The second one is that the style is not included anywhere, so it was never getting used. To fix that you will need to add the following child to the `Application.Styles` tag in App.axaml

```xml
<StyleInclude Source="/Controls/TestTemplatedControl.axaml" />
```

After those 2 steps, everything worked perfectly.

**Reply by FraserElectronics** [reply]
Thank you very much for that. It was an oversight not to add the StyleInclude, I have a number of templates controls so know I have to include them but it was the binding I could not figure out.

Thanks again.

Andy
